File: source/winx86/NetLayer/IpHandler.cs
using System;
using System.Collections.Generic;
using System.Text;
using System.Threading;
using System.Net;

namespace NetLayer
{
    /// <summary>
    /// Clase que enumera los estados posibles de IP 
    /// </summary>
    internal static class IpState
    {
        public const int NOTFOUND = 0;
        public const int INVALID = 1;
        public const int DUPLICATE = 2;
        public const int VALID = 3;
    }

    /// <summary>
    /// Clase que enumera los estados posibles de la conexión inalambrica
    /// </summary>
    internal static class IphandlerState
    {
        public const Int32 STARTEDSTRONG = 1;
        public const Int32 STOPPED = 2;
        public const Int32 STARTEDWEAK = 3;
    }
    
    /// <summary>
    /// Clase que se preocupa de que no hayan conflictos con la IP
    /// </summary>
    internal class IpHandler
    {
        /// <summary>
        /// Declaración de delegado para evento de cambio de ip
        /// </summary>
        public delegate void ResetIpDelegate();
        /// <summary>
        /// Handler para cambio de ip
        /// </summary>
        /// <returns>vacio</returns>
        private ResetIpDelegate resetIpDelegate;
        /// <summary>
        /// Thread que verifica la IP
        /// </summary>
        private Thread checkIpThread;
        /// <summary>
        /// El estado de este objeto (un parametro de IpHandlerState)
        /// </summary>
        private Int32 state;
        /// <summary>
        /// Objeto de lock
        /// </summary>
        private Object stopLock;
        /// <summary>
        /// Cola de Ip's de mensajes i'm alive recibidos
        /// </summary>
        private Queue<IPAddress> queue;
        /// <summary>
        /// Objeto de lock para la cola
        /// </summary>
        private Object queueLock;
        /// <summary>
        /// Los datos de red
        /// </summary>
        private NetData netData;
        /// <summary>
        /// Contador de loliness
        /// </summary>
        private Int32 lolinessTimeOut;
        /// <summary>
        /// valor que cambia cuando el usuario esta correctamente difundiendo mensajes
        /// </summary>
        private Int64 lastAliveValue;

        /// <summary>
        /// Constructor
        /// </summary>
        /// <param name="netData">Los datos de configuración de red</param>
        /// <param name="resetIpDelegate">la función que se gatilla cuando la Ip debe ser cambiada</param>
        public IpHandler(NetData netData, ResetIpDelegate resetIpDelegate)
        {
            this.resetIpDelegate = resetIpDelegate;
            this.netData = netData;
            checkIpThread = new Thread(new ThreadStart(checkIp));
            state = IphandlerState.STOPPED;
            stopLock = new Object();
            queueLock = new Object();
            queue = new Queue<IPAddress>();
            lolinessTimeOut = 0;
            lastAliveValue = 0;
        }

        /// <summary>
        /// Comienza la verificación fuerte
        /// </summary>
        public void startStrongDAD()
        {
            state = IphandlerState.STARTEDSTRONG;
            checkIpThread.Start();
        }

        /// <summary>
        /// Comienza la verificación débil
        /// </summary>
        public void chageToWeakDAD()
        {
            state = IphandlerState.STARTEDWEAK;
        }

        /// <summary>
        /// Detiene la verificación
        /// </summary>
        public void stop()
        {
            lock (stopLock)
            {
                state = IphandlerState.STOPPED;
            }
        }

        /// <summary>
        /// Verifica que no exista Ip duplicada en el sistema operativo, si gatilla resetIpDelegate, asegura que el Thread se detendrá
        /// </summary>
        private void checkIp()
        {
            while (true)
            {
                System.Threading.Thread.Sleep(netData.WaitTimeStart);
                //Chequea Strong DAD
                if (state.Equals(IphandlerState.STARTEDSTRONG))
                {
                    int ipState = SystemHandler.getIpState(netData.NetworkAdapter, netData.IpTcpListener);
                    switch (ipState)
                    {
                        case IpState.DUPLICATE:
                            {
                                resetIpDelegate();
                                return;
                            }
                        case IpState.NOTFOUND:
                            {
                                resetIpDelegate();
                                return;
                            }
                    }
                }
                else if (state.Equals(IphandlerState.STARTEDWEAK))
                {
                    //Chequea Weak DAD
                    lock (queueLock)
                    {
                        while (queue.Count > 0)
                        {
                            IPAddress outterIp = queue.Dequeue();
                            if (outterIp.Equals(netData.IpTcpListener))
                            {
                                resetIpDelegate();
                                return;
                            }
                        }
                    }
                    //chequea Strong DAD
                    int ipState = SystemHandler.getIpState(netData.NetworkAdapter, netData.IpTcpListener);
                    switch (ipState)
                    {
                        case IpState.DUPLICATE:
                            {
                                resetIpDelegate();
                                return;
                            }
                        case IpState.INVALID:
                            {
                                resetIpDelegate();
                                return;
                            }
                        case IpState.NOTFOUND:
                            {
                                resetIpDelegate();
                                return;
                            }
                    }
                    //Chequea loneliness
                    if (netData.OpSystem == OpSystemType.WIN7)
                    {
                        Int64 nBytes = SystemHandler.getAliveValue(netData.NetworkAdapter);
                        if (nBytes == lastAliveValue)
                        {
                            lolinessTimeOut++;
                            if (lolinessTimeOut >= netData.LolinessTimeOut)
                            {
                                resetIpDelegate();
                                return;
                            }
                        }
                        else
                        {
                            lolinessTimeOut = 0;
                        }
                        lastAliveValue = nBytes;
                    }

                    //Chequea operatividad del adaptador
                    if (netData.OpSystem == OpSystemType.WIN7)
                    {
                        if (!SystemHandler.isOperative(netData.NetworkAdapter))
                        {
                            resetIpDelegate();
                            return;
                        }
                    } 

                    
                }
                //Detiene o duerme segun corresponda
                lock (stopLock)
                {
                    if (state.Equals(IphandlerState.STOPPED))
                    {
                        return;
                    }
                }
                
            }
        }

        /// <summary>
        /// Coloca una ip a la cola de ips para chequear
        /// </summary>
        /// <param name="outterIp">la ip a encolar</param>
        public void put(IPAddress outterIp)
        {
            lock (queueLock)
            {
                queue.Enqueue(outterIp);
            }
        }
    }
}

File: source/java/hlmp/NetLayer/IpHandler.java
package hlmp.NetLayer;

import java.util.concurrent.ConcurrentLinkedQueue;
import hlmp.NetLayer.Constants.*;
import hlmp.NetLayer.Interfaces.ResetIpHandler;
import hlmp.NetLayer.Interfaces.WifiHandler;

public class IpHandler{

	/**
	 * Handler para cambio de ip
	 */
	private ResetIpHandler resetIpHandler;

	/**
	 * Thread que verifica la IP
	 */
    private Thread checkIpThread;
	/**
	 * El estado de este objeto (un parametro de IpHandlerState)
	 */
	private int state;
	/**
	 * Objeto de lock
	 */
	private Object stopLock;
	/**
	 * Cola de Ip's de mensajes i'm alive recibidos
	 */
	private ConcurrentLinkedQueue<String> queue;
	/**
	 * Objeto de lock para la cola
	 */
	private Object queueLock;
	/**
	 * Los datos de red
	 */
	private NetData netData;

	/**
	 * Contador de loliness
	 */
	//private int lolinessTimeOut;
	/**
	 * valor que que cambia si el usuario esta correctamente recibiendo la multidifusión
	 */

	private WifiHandler wifiHandler;
	/**
	 * valor que cambia cuando el usuario esta correctamente difundiendo mensajes
	 */
	//private int lastAliveValue;


	public IpHandler(NetData netData , ResetIpHandler resetIpHandler, WifiHandler wifiHandler) {
		this.wifiHandler = wifiHandler;
		this.resetIpHandler = resetIpHandler;
		this.netData = netData;
		this.checkIpThread = getCheckIpThread();
		this.checkIpThread.setName("IpHandler Thread");
		this.state = IphandlerState.STOPPED;
		this.stopLock = new Object();
		this.queueLock = new Object();
		queue = new ConcurrentLinkedQueue<String>();
		//this.lolinessTimeOut = 0;
		//this.lastAliveValue = 0;

	}

	/**
	 * Comienza la verificación fuerte
	 */
	public void startStrongDAD(){
		this.state = IphandlerState.STARTEDSTRONG;
		this.checkIpThread.start();
	}

	/**
	 * Comienza la verificación débil
	 */
	public void chageToWeakDAD(){
		this.state = IphandlerState.STARTEDWEAK;
	}

	/**
	 * Detiene la verificación
	 */
	public void stop(){
		
		synchronized (stopLock){
			this.state = IphandlerState.STOPPED;
		}
		try {
			checkIpThread.interrupt();
			checkIpThread.join();
		} catch (InterruptedException e) {
			return;
		}
	}
	
	/**
	 * Verifica que no exista Ip duplicada en el sistema operativo, si gatilla resetIpDelegate, asegura que el Thread se detendrá
	 */
	private Thread getCheckIpThread(){
		return new Thread(){

			public void run(){
				// TODO: fvalverd corroborar que getCheckIpThread funciona
				while(true){
					try {
						sleep(netData.getWaitTimeStart());
					} catch (InterruptedException e) {
						return;
					}
					//Chequea Strong DAD
					if (state == IphandlerState.STARTEDSTRONG)
					{
						int ipState = wifiHandler.getIpState();
						switch (ipState)
						{
						case IpState.DUPLICATE:
						{
							resetIpHandler.resetIp();
							return;
						}
						case IpState.NOTFOUND:
						{
							resetIpHandler.resetIp();
							return;
						}
						}
					}
					else if (state == IphandlerState.STARTEDWEAK)
					{
						//Chequea Weak DAD
						synchronized (queueLock){
							while (queue.size() > 0){
								String outterIp = queue.poll();
								if (outterIp.equals(netData.getIpTcpListener())){
									resetIpHandler.resetIp();
									return;
								}
							}
						}
						//chequea Strong DAD
						switch (wifiHandler.getIpState()) {
							case IpState.DUPLICATE:
							{
								resetIpHandler.resetIp();
								return;
							}
							case IpState.INVALID:
							{
								resetIpHandler.resetIp();
								return;
							}
							case IpState.NOTFOUND:
							{
								resetIpHandler.resetIp();
								return;
							}
						}
						
						// Detiene o duerme segun corresponda
						synchronized (stopLock){
							if (state == IphandlerState.STOPPED){
								return;
							}
						}

					}
				}
			}
		};
	}
	
	/**
	 * Coloca una ip a la cola de ips para chequear
	 * @param outterIp la ip a encolar
	 */
	public void put(String outterIp)
	{
		synchronized (queueLock){
			this.queue.add(outterIp);
		}
	}
	
}